File: Problem1_ProductExceptSelf.cs
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Using Running product approach get the running products of the array numbers from left to right
// and right to left. The product of each element of rProd1 with corresponding element of rProd2 gives the result.
public class Solution {
    public int[] ProductExceptSelf(int[] nums) {
        int[] rProd1 = new int[nums.Length];
        rProd1[0] = 1;
        // Left to Right
        for(int i = 1; i < nums.Length; i++) {
            rProd1[i] = rProd1[i-1] * nums[i-1];
        }
        
        //Right to Left
         int[] rProd2 = new int[nums.Length];
        rProd2[nums.Length - 1] = 1;
        for(int i = nums.Length - 2; i >= 0; i--) {
            rProd2[i] = rProd2[i+1] * nums[i+1];
        }

        int[] result = new int[nums.Length];
        //Result -> product of rProd1 & rProd2 with corresponding indexed elements
        for(int i = 0; i < nums.Length; i++) {
            result[i] = rProd1[i] * rProd2[i];
        }

        return result;
    }
}

File: Problem-1-Product-Except-Self.java
// Time Complexity :O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
/* Three line explanation of solution in plain english : We are taking resultant product (rp) and traversing left to right and getting all the left products and right to left and getting all the right products except the num itself. And finally we are mutiplying both the products. All this is stored in a result array */

// Your code here along with comments explaining your approach
class Solution {
    public int[] productExceptSelf(int[] nums) {
		
	int result[] = new int[nums.length];//result array

	int rp=1;//resultant product (rp) 
	
	for( int i=0; i < nums.length; i++){
	
	result[i]=rp;// left products
	rp=rp*nums[i]; 
	
	}
     
	rp = 1;
	for( int i=nums.length-1; i>=0 ; i--){
	
	result[i]=rp*result[i]; // right products
	rp=rp*nums[i];
	
	}
	
	return result;
	 
    }
}
